app/gui2/shared/ast/parse.ts

```diff
@@ -591,11 +596,121 @@ function resync(
     const goodAst = goodSpans.get(span)?.[0]
     // The parent of the root of a bad subtree must be a good AST.
     assertDefined(goodAst)
-    parent.replaceValue(edit.copy(goodAst))
+    parent.syncToCode(goodAst.code())
   }
 
   console.warn(
     `repair: Replaced ${parentsOfBadSubtrees.size} subtrees with their reparsed equivalents.`,
     parentsOfBadSubtrees,
   )
 }
+
+function hashString(s: string) {
+  return hash(s)
+}
```

Comment:
I don't quite see what do we get from this function.

Reply:
`hash` is very general and will accept non-strings, which is not intended, so this adds some type safety.

Reply:
In the next PR I used xxhash128. 128-bit might be excessive, but this will never be a performance bottleneck.